### ISSUE TITLE
Proposal: add `Close()` method to `index.Postings` interface

### ISSUE THREAD

**Issue by colega** [prometheus/prometheus#15160]
### Proposal

Under heavy loads we allocate lots of `ListPostings` and similar objects, and we could benefit from some pooling to reduce pressure on the GC.

However, there's no clear closing call on the postings, even though the contract is clear: once `Postings.Next()` is false, there's nothing else we can do with them.

What if we added a `Close()` method to postings interface and did some recycling in that call, assuming that the instance isn't going to be used anymore.

The largest issue I see is that existing codebases (especially downstream projects) will fail to review all their calls, so we won't be able to rely on that call for a while, and we'll have to be just optimistic.

An alternative I considered was to implement just a `recycle()` method on `ListPostings` and `bigEndianPostings`, and call it from `intersectPostings` and `removedPostings`, as those are likely the major consumers of the former ones.

**Comment by krajorama** [comment]
Hello from the bug scrub!

Not against it on principle, we welcome PRs in this area.

**Comment by Sabadon** [comment]
Hello, I’ve looked into this issue a bit and would like to share my understanding of what needs to be done, so I can get feedback on whether I’m heading in the right direction. We would add a pool that holds `ListPostings` objects, which would be retrieved from the pool when `newListPostings` is called, and then, when `recycle()` is called, they would be cleared and returned back to the pool, right?
Here is some example code:
```golang
const initialCapacity = 128
var pool = sync.Pool {
  New: func() any {
    return &ListPostings{
      list: make([]storage.SeriesRef, 0, initialCapacity)
    }
  }
}

type ListPostings struct {
	list []storage.SeriesRef
	cur  storage.SeriesRef
}

func NewListPostings(list []storage.SeriesRef) Postings {
	return newListPostings(list...)
}

func newListPostings(list ...storage.SeriesRef) *ListPostings {
  p := pool.Get().(*ListPostings)

  p.list = p.list[:0]
  p.cur = 0

  p.list = append(p.list, list...)
  return p
}

func (p *ListPostings) recycle() {
  p.list[:0]
  p.cur = 0

  pool.Put(p)
}

```

**Comment by dimitarvdimitrov** [comment]
yes, that's my understanding of it too @Sabadon ; i would go with the Close implementation though. `Close` gives us the same benefits that `recycle` does (internal calling within the `index` package) + allows clients of the package to also call `Close` and return postings to the pool
